pkg/provisioners/heartbeats/controller/reconcile.go

```diff
@@ -0,0 +1,263 @@
+		return reconcile.Result{}, nil
+	}
+
+	if err != nil {
+		logger.Errorf("could not fetch Source %v for %+v\n", err, request)
+		return reconcile.Result{}, err
+	}
+
+	if source.Spec.Provisioner.Ref.Name != provisionerName {
+		logger.Errorf("heartbeats skipping source %s, provisioned by %s\n", source.Name, source.Spec.Provisioner.Ref.Name)
+		return reconcile.Result{}, nil
+	}
+
+	original := source.DeepCopy()
+
+	// Reconcile this copy of the Source and then write back any status
+	// updates regardless of whether the reconcile error out.
+	err = r.reconcile(ctx, source)
+	if err != nil {
+		logger.Warnf("Failed to reconcile source: %v", err)
+	}
+	if equality.Semantic.DeepEqual(original.Status, source.Status) {
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+	} else if _, err := r.updateStatus(source); err != nil {
+		logger.Warnf("Failed to update source status: %v", err)
+		return reconcile.Result{}, err
+	}
+
+	// Requeue if the resource is not ready:
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) reconcile(ctx context.Context, source *v1alpha1.Source) error {
+	logger := logging.FromContext(ctx)
+
+	source.Status.InitializeConditions()
+
+	// TODO:
+	//// See if the source has been deleted
+	//accessor, err := meta.Accessor(source)
+	//if err != nil {
+	//	logger.Warnf("Failed to get metadata accessor: %s", err)
+	//	return err
+	//}
+	//deletionTimestamp := accessor.GetDeletionTimestamp()
+
+	args := &resources.HeartBeatArguments{}
+
+	if source.Spec.Arguments != nil {
+		if err := json.Unmarshal(source.Spec.Arguments.Raw, args); err != nil {
+			logger.Infof("Error: %s failed to unmarshal arguments, %v", source.Name, err)
+		}
+	}
+	args.Name = source.Name
+	args.Namespace = source.Namespace
+
+	channel, err := r.getChannel(ctx, source)
+	if err != nil {
+		fqn := "Channel.eventing.knative.dev/v1alpha1"
+		if errors.IsNotFound(err) {
+			channel, err = r.createChannel(ctx, source, nil, args)
+			if err != nil {
+				return err
+			}
+			r.recorder.Eventf(source, corev1.EventTypeNormal, "Provisioned", "Created channel %q", channel.Name)
+			source.Status.SetProvisionedObjectState(channel.Name, fqn, "Created", "Created channel %q", channel.Name)
+			source.Status.MarkDeprovisioned("Provisioning", "Provisioning Channel %s", args.Name)
+		} else {
+			if channel.Status.IsReady() {
+				source.Status.SetProvisionedObjectState(channel.Name, fqn, "Ready", "")
+			}
+		}
+
+		source.Status.Subscribable.Channelable.Namespace = channel.Namespace
+		source.Status.Subscribable.Channelable.Name = channel.Name
+		source.Status.Subscribable.Channelable.APIVersion = "eventing.knative.dev/v1alpha1"
+		source.Status.Subscribable.Channelable.Kind = "Channel"
+
+	}
+
+	pod, err := r.getPod(ctx, source)
```

Comment:
A Deployment would offer better semantics around updates, which are not handled today.

Reply:
Moved to a deployment.